### PR TITLE
Add change group chat name endpoint

### PR DESCRIPTION
**Pull request by cammellos**
Add an endpoint to change group chat name and allow leaving without removing the chat.

### CODE REVIEW

File: protocol/messenger.go [hunk 1]
```diff
@@ -823,6 +823,70 @@ func (m *Messenger) AddMembersToGroupChat(ctx context.Context, chatID string, me
 	return &response, m.saveChat(chat)
 }
 
+func (m *Messenger) ChangeGroupChatName(ctx context.Context, chatID string, name string) (*MessengerResponse, error) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
```
Comment (adambabik) [inline protocol/messenger.go:830]: The lock's scope is too large. For example, `event.Sign` can take significant time and it does not make sense to block for that long any other operations.

I see it's in other methods like this. We can keep it for consistency and fix in a different PR.

File: protocol/messenger.go [hunk 2]
```diff
@@ -823,6 +823,70 @@ func (m *Messenger) AddMembersToGroupChat(ctx context.Context, chatID string, me
...
+
+	var response MessengerResponse
```
Comment (adambabik) [inline protocol/messenger.go]: We can move it down to where it is actually used.

File: protocol/messenger.go [hunk 3]
```diff
@@ -823,6 +823,70 @@ func (m *Messenger) AddMembersToGroupChat(ctx context.Context, chatID string, me
...
+	logger := m.logger.With(zap.String("site", "ChangeGroupChatName"))
```
Comment (adambabik) [inline protocol/messenger.go]: We can move it to the top of the method in order not to clutter the logic.

File: protocol/messenger.go [hunk 4]
```diff
@@ -823,6 +823,70 @@ func (m *Messenger) AddMembersToGroupChat(ctx context.Context, chatID string, me
...
+	logger.Info("Changing group chat name", zap.String("chatID", chatID), zap.String("name", name))
+	chat, ok := m.allChats[chatID]
+	if !ok {
+		return nil, errors.New("can't find chat")
```
Comment (adambabik) [inline protocol/messenger.go]: `errors.New("can't find chat")` can be defined as a variable.

File: protocol/messenger.go [hunk 5]
```diff
@@ -823,6 +823,70 @@ func (m *Messenger) AddMembersToGroupChat(ctx context.Context, chatID string, me
 	return &response, m.saveChat(chat)
 }
 
+func (m *Messenger) ChangeGroupChatName(ctx context.Context, chatID string, name string) (*MessengerResponse, error) {
```
Comment (adambabik) [inline protocol/messenger.go:825]: No unit test?

**Review by adambabik — COMMENTED** [review]
> and allow leaving without removing the chat.

What will be the consequences?

**Comment by cammellos** [comment]
@adambabik addressed feedback
> What will be the consequences?

None, the user will still be able to read old messages, won't receive new ones

**Review by adambabik — APPROVED** [review]

**Review by flexsurfer — APPROVED** [review]

File: protocol/messenger.go [hunk 6]
```diff
@@ -821,6 +822,71 @@ func (m *Messenger) AddMembersToGroupChat(ctx context.Context, chatID string, me
...
+	}
+
+	encodedMessage, err := m.processor.EncodeMembershipUpdate(group, nil)
+	if err != nil {
+		return nil, err
+	}
+	_, err = m.dispatchMessage(ctx, &RawMessage{
+		LocalChatID:         chat.ID,
+		Payload:             encodedMessage,
+		MessageType:         protobuf.ApplicationMetadataMessage_MEMBERSHIP_UPDATE_MESSAGE,
```
Comment (jakubgs) [inline protocol/messenger.go:869]: Maybe renaming this event to something like `GROUP_STATE_CHANGE` would make it clearer as to what it does.

File: protocol/messenger.go [hunk 7]
```diff
@@ -821,6 +822,71 @@ func (m *Messenger) AddMembersToGroupChat(ctx context.Context, chatID string, me
...
+	clock, _ := chat.NextClockAndTimestamp(m.getTimesource())
+	// Add members
+	event := v1protocol.NewNameChangedEvent(name, clock)
+	event.ChatID = chat.ID
+	err = event.Sign(m.identity)
+	if err != nil {
+		return nil, err
+	}
+
+	err = group.ProcessEvent(event)
```
Comment (jakubgs) [inline protocol/messenger.go:852]: A comment saying something like `// Update in-memory stuctures` would be nice.

File: protocol/messenger.go [hunk 8]
```diff
@@ -821,6 +822,71 @@ func (m *Messenger) AddMembersToGroupChat(ctx context.Context, chatID string, me
...
+	if err != nil {
+		return nil, err
+	}
+
+	recipients, err := stringSliceToPublicKeys(group.Members(), true)
+	if err != nil {
+		return nil, err
+	}
+
+	encodedMessage, err := m.processor.EncodeMembershipUpdate(group, nil)
```
Comment (jakubgs) [inline protocol/messenger.go:862]: A comment explaining that this encodes the whole state of thee group to be sent, not just name change.

File: protocol/messenger.go [hunk 9]
```diff
@@ -953,15 +1019,15 @@ func (m *Messenger) ConfirmJoiningGroup(ctx context.Context, chatID string) (*Me
 	return &response, m.saveChat(chat)
 }
 
-func (m *Messenger) LeaveGroupChat(ctx context.Context, chatID string) (*MessengerResponse, error) {
+func (m *Messenger) LeaveGroupChat(ctx context.Context, chatID string, remove bool) (*MessengerResponse, error) {
```
Comment (jakubgs) [inline protocol/messenger.go:1023]: This change should probably be a separate commit/pr.

File: protocol/messenger_test.go [hunk 10]
```diff
@@ -874,6 +874,67 @@ func (s *MessengerSuite) TestRetrieveTheirPrivateGroupChat() {
 	s.Require().NotNil(actualChat.LastMessage)
 }
 
+// Test receiving a message on an existing private group chat
+func (s *MessengerSuite) TestChangeNameGroupChat() {
+	var response *MessengerResponse
+	theirMessenger := s.newMessenger(s.shh)
+	response, err := s.m.CreateGroupChatWithMembers(context.Background(), "id", []string{})
```
Comment (jakubgs) [inline protocol/messenger_test.go]: Maybe rename `"id"` to `"old-name"` to make it clear that second argument is the group chat name.

File: protocol/messenger_test.go [hunk 11]
```diff
@@ -874,6 +874,67 @@ func (s *MessengerSuite) TestRetrieveTheirPrivateGroupChat() {
...
+	// Retrieve their messages so that the chat is created
+	err = tt.RetryWithBackOff(func() error {
+		var err error
+		response, err = theirMessenger.RetrieveAll()
+		if err == nil && len(response.Chats) == 0 {
+			err = errors.New("chat invitation not received")
+		}
+		return err
+	})
+	s.Require().NoError(err)
```
Comment (jakubgs) [inline protocol/messenger_test.go:903]: What's the difference between `s.NoError(err)` and `s.Require().NoError(err)`?

File: protocol/messenger_test.go [hunk 12]
```diff
@@ -874,6 +874,67 @@ func (s *MessengerSuite) TestRetrieveTheirPrivateGroupChat() {
...
+
+	_, err = theirMessenger.ConfirmJoiningGroup(context.Background(), ourChat.ID)
+	s.NoError(err)
+
+	err = tt.RetryWithBackOff(func() error {
```
Comment (jakubgs) [inline protocol/messenger_test.go:909]: Add a comment as above like `// Retrieve messages to verify group was joined`. Maybe this could be a helper method called `retrieveOrError()` or something like that. Since we do it three times.

File: protocol/messenger_test.go (hunk 11 above)
Comment (cammellos) [reply protocol/messenger_test.go:903]: `s.NoError` will continue the test, while `require().NoError()` will stop execution, which is what probably you want most of the times
